File: csharp/day12_part1_2020.cs

using System;
using System.IO;

class Ship
{
    public int x, y, facing;

    public void ProcessInstruction(char action, int value)
    {
        switch (action)
        {
            case 'N':
                y += value;
                break;
            case 'S':
                y -= value;
                break;
            case 'E':
                x += value;
                break;
            case 'W':
                x -= value;
                break;
            case 'L':
                facing = (facing - value + 360) % 360;
                break;
            case 'R':
                facing = (facing + value) % 360;
                break;
            case 'F':
                switch (facing)
                {
                    case 0:
                        x += value;
                        break;
                    case 90:
                        y -= value;
                        break;
                    case 180:
                        x -= value;
                        break;
                    case 270:
                        y += value;
                        break;
                }
                break;
        }
    }
}

class Program
{
    static void Main()
    {
        using (StreamReader sr = new StreamReader("input.txt"))
        {
            Ship ship = new Ship { x = 0, y = 0, facing = 0 };
            string line;
            while ((line = sr.ReadLine()) != null)
            {
                char action = line[0];
                int value = int.Parse(line.Substring(1));
                ship.ProcessInstruction(action, value);
            }

            int manhattanDistance = Math.Abs(ship.x) + Math.Abs(ship.y);
            Console.WriteLine(manhattanDistance);
        }
    }
}

File: swift/day12_part1_2020.swift
import Foundation

struct Ship {
    var x, y, facing: Int

    mutating func processInstruction(_ action: Character, _ value: Int) {
        switch action {
        case "N":
            y += value
        case "S":
            y -= value
        case "E":
            x += value
        case "W":
            x -= value
        case "L":
            facing = (facing - value + 360) % 360
        case "R":
            facing = (facing + value) % 360
        case "F":
            switch facing {
            case 0:
                x += value
            case 90:
                y -= value
            case 180:
                x -= value
            case 270:
                y += value
            default:
                break
            }
        default:
            print("Invalid action: \(action)")
        }
    }
}

do {
    let fileURL = URL(fileURLWithPath: "input.txt")
    let content = try String(contentsOf: fileURL, encoding: .utf8)
    let lines = content.components(separatedBy: "\n")

    var ship = Ship(x: 0, y: 0, facing: 0)
    for line in lines {
        let action = line.first!
        let value = Int(String(line.dropFirst()))!
        ship.processInstruction(action, value)
    }

    let manhattanDistance = abs(ship.x) + abs(ship.y)
    print(manhattanDistance)
} catch {
    print("Error reading file:", error)
}